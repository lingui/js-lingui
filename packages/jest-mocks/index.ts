export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
