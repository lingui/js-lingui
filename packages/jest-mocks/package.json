{
  "private": true,
  "name": "@lingui/jest-mocks",
  "version": "0.0.0-managed-by-release-script",
  "description": "Mocks for Jest",
  "main": "index.js",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "license": "MIT",
  "keywords": [
    "jest",
    "testing",
    "mock"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/lingui/js-lingui.git"
  },
  "bugs": {
    "url": "https://github.com/lingui/js-lingui/issues"
  },
  "engines": {
    "node": ">=6.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js",
    "*.js.flow"
  ]
}
