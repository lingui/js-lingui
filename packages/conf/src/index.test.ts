import path from "path"
import { validate } from "jest-validate"
import {
  getConfig,
  replaceRootDir,
  catalogMigration,
  configValidation,
} from "@lingui/conf"
import { mockConsole, getConsoleMockCalls } from "@lingui/jest-mocks"

describe("@lingui/conf", function () {
  it("should return default config", function () {
    expect.assertions(2)

    mockConsole((console) => {
      const config = getConfig({
        cwd: path.resolve(__dirname, path.join("fixtures", "valid")),
      })
      expect(console.warn).not.toBeCalled()
      expect(config).toMatchSnapshot()
    })
  })

  it("should validate `locale`", function () {
    expect.assertions(2)

    mockConsole((console) => {
      getConfig()
      expect(console.warn).not.toBeCalled()
      expect(console.error).toBeCalledWith(
        expect.stringContaining("No locales defined")
      )
    })
  })

  it("should replace <rootDir>", function () {
    const config = replaceRootDir(
      // @ts-ignore
      {
        compileNamespace: "cjs",
        catalogs: [
          {
            path: "/",
            include: ["<rootDir>/src"],
            exclude: ["<rootDir>/ignored"],
          },
        ],
      },
      "/Root"
    )

    expect(config.compileNamespace).toEqual("cjs")
    expect(config.catalogs).toEqual([
      {
        path: "/",
        include: ["/Root/src"],
        exclude: ["/Root/ignored"],
      },
    ])
  })

  describe("catalogMigration", function () {
    it("should show deprecation warning", function () {
      mockConsole((console) => {
        validate(
          {
            localeDir: "./locale",
            srcPathDirs: ["./src"],
            srcPathIgnorePatterns: ["/node_modules/"],
          },
          configValidation
        )
        expect(getConsoleMockCalls(console.warn)).toMatchSnapshot()
      })
    })

    it("shouldn't provide default config if no obsolete config is defined", function () {
      const config = {}

      // @ts-ignore
      expect(catalogMigration(config)).toEqual({})
    })

    it("should normalize string localeDir", function () {
      const config = {
        localeDir: "./locales",
      }

      // @ts-ignore
      expect(catalogMigration(config)).toEqual({
        catalogs: [
          {
            path: "locales/{locale}/messages",
            include: ["<rootDir>"],
            exclude: ["**/node_modules/**"],
          },
        ],
      })
    })

    it("should migrate srcPathDirs and srcPathIgnorePatterns", function () {
      const config = {
        localeDir: "locales",
        srcPathDirs: ["src"],
        srcPathIgnorePatterns: ["src/node_modules/"],
      }

      // @ts-ignore
      expect(catalogMigration(config)).toEqual({
        catalogs: [
          {
            path: "locales/{locale}/messages",
            include: ["src"],
            exclude: ["src/node_modules/"],
          },
        ],
      })
    })
  })

  it("searches for a config file", function () {
    // hide validation warning about missing locales
    mockConsole(() => {
      const config = getConfig({
        cwd: path.resolve(__dirname, path.join("fixtures", "valid")),
      })
      expect(config.locales).toEqual(["en-gb"])
    })
  })

  describe("with configPath parameter", function () {
    it("allows specific config file to be loaded", function () {
      // hide validation warning about missing locales
      mockConsole(() => {
        const config = getConfig({
          configPath: path.resolve(
            __dirname,
            path.join("fixtures", "valid", "custom.config.js")
          ),
        })
        expect(config.locales).toEqual(["cs", "sk"])
      })
    })
  })
})
