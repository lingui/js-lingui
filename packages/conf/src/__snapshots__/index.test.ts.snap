// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@lingui/conf catalogMigration should show deprecation warning 1`] = `
● Deprecation Warning:

 Option localeDir is deprecated. Configure source paths using catalogs instead.

    @lingui/cli now treats your current configuration as:

    {
      "catalogs": [
  {
    "path": "locale/{locale}/messages",
    "include": [
      "./src"
    ],
    "exclude": [
      "/node_modules/"
    ]
  }
]
    }

    Please update your configuration.
    

Documentation: https://lingui.js.org/ref/conf.html
● Deprecation Warning:

 Option srcPathDirs is deprecated. Configure source paths using catalogs instead.

    @lingui/cli now treats your current configuration as:

    {
      "catalogs": [
  {
    "path": "locale/{locale}/messages",
    "include": [
      "./src"
    ],
    "exclude": [
      "/node_modules/"
    ]
  }
]
    }

    Please update your configuration.
    

Documentation: https://lingui.js.org/ref/conf.html
● Deprecation Warning:

 Option srcPathIgnorePatterns is deprecated. Configure excluded source paths using catalogs instead.

    @lingui/cli now treats your current configuration as:

    {
      "catalogs": [
  {
    "path": "locale/{locale}/messages",
    "include": [
      "./src"
    ],
    "exclude": [
      "/node_modules/"
    ]
  }
]
    }

    Please update your configuration.
    

Documentation: https://lingui.js.org/ref/conf.html
`;

exports[`@lingui/conf should return default config 1`] = `
Object {
  catalogs: Array [
    Object {
      exclude: Array [
        */node_modules/*,
      ],
      include: Array [
        .,
      ],
      path: ./locale/{locale}/messages,
    },
  ],
  compileNamespace: cjs,
  extractBabelOptions: Object {
    plugins: Array [],
    presets: Array [],
  },
  fallbackLocale: ,
  format: po,
  formatOptions: Object {
    origins: true,
  },
  locales: Array [
    en-gb,
  ],
  orderBy: messageId,
  pseudoLocale: ,
  rootDir: .,
  runtimeConfigModule: Array [
    @lingui/core,
    i18n,
  ],
  sourceLocale: ,
}
`;
